feat: FIT-173: Audio spectrograms

cloudmark opened the pull request:
# Spectrogram visualization to Audio Component

## Reason for change
This PR adds spectrogram visualization support to the audio editor, enabling users to visualize frequency content over time in audio recordings. This feature enhances audio annotation capabilities by providing visual frequency analysis tools, particularly useful for tasks like speech analysis, music transcription, and sound event detection.

The implementation includes:
- FFT-based spectrogram rendering with configurable parameters
- Multiple color scheme options with live preview
- Mel-scale frequency mapping support
- Real-time parameter adjustment
- Performance optimizations for smooth rendering
- Modular code organization for maintainability

## Screenshots

1. **Configuration Interface**
<img width="1763" alt="Screenshot 2025-04-20 at 19 05 32" src="https://github.com/user-attachments/assets/ffdd5fa5-a354-4a23-bc75-c937b670ad97" />

Shows the labeling interface configuration with the new `spectrogram="true"` property in the XML configuration, demonstrating how the feature can be enabled through the labeling interface.

2. **Color Scheme Selection**
<img width="1763" alt="Screenshot 2025-04-20 at 19 05 01" src="https://github.com/user-attachments/assets/4d0bc606-8d99-4aed-8e62-02a5638686b7" />

Demonstrates the color scheme selection interface with:
- Multiple predefined schemes (Autumn, Bathymetry, Blackbody, etc.)
- Visual preview for each scheme
- Real-time application of color changes
- Smooth transition between schemes

3. **Time Display on Hover**
<img width="1763" alt="Screenshot 2025-04-20 at 19 04 50" src="https://github.com/user-attachments/assets/8d6a532d-ec11-4c5c-ac3c-5e17456503e9" />

Shows interactive features:
- Precise time indicator (00:00:00.447) on hover
- Clear visualization of frequency content
- Smooth rendering of the spectrogram
- Integration with the waveform display

4. **Playback and Spectrogram Settings**
<img width="1763" alt="Screenshot 2025-04-20 at 19 04 47" src="https://github.com/user-attachments/assets/735f0dc3-d0eb-4b20-bcfa-710527bb9008" />

Comprehensive control panel featuring:
- Playback speed adjustment
- Audio zoom y-axis control
- FFT Samples slider (64-2048)
- Loop Regions toggle
- Auto-play New Regions toggle
- Integrated spectrogram controls

5. **Advanced Spectrogram Controls**
<img width="1763" alt="Screenshot 2025-04-20 at 19 04 43" src="https://github.com/user-attachments/assets/a4c05e2b-fb83-4372-8f18-051890f72dfd" />

Detailed configuration options:
- FFT Samples selection (512)
- Mel Bands adjustment (64)
- Spectrogram dB range (-80 to -10)
- Windowing Function selection (Blackman)
- Color Scheme selection (Viridis)
- View toggles for timeline, audio wave, and spectrogram

## Rollout strategy
The feature is implemented with a progressive enhancement approach:

1. **Feature Flag**: 
```xml
<AudioPlus name="audio"
          value="$audio"
          hotkey="space"
          sync="group_a"
          defaultscale="1"
          defaultzoom="2"
          zoom="true"
          spectrogram="true"
/>
```

2. **Backward Compatibility**: 
- Existing audio components continue to work without changes
- Spectrogram can be enabled/disabled per instance
- All new parameters have sensible defaults

3. **Performance Considerations**:
- Lazy loading of FFT computation code
- Progressive rendering for large files
- Configurable quality settings

## Testing
Comprehensive testing strategy:

1. **Unit Tests**:
- WindowFunctions.ts: Window function calculations
- ColorMapper.ts: Color scheme management
- FFT computation accuracy
- Parameter validation

2. **Integration Tests**:
- Audio loading and visualization
- Real-time parameter updates
- Color scheme switching
- Performance benchmarks

3. **Manual Testing Scenarios**:
- Various audio formats (WAV, MP3, OGG)
- Different file lengths (short clips to long recordings)
- Multiple sample rates and bit depths
- Browser compatibility (Chrome, Firefox, Safari)
- Performance with large files

## Risks
1. **Performance**:
- FFT computation is CPU-intensive
- Mitigated through:
  - Chunked rendering
  - Yield scheduling
  - Canvas optimization
  - Caching mechanisms

2. **Memory Usage**:
- Large audio files require more memory for FFT
- Mitigated through:
  - Buffer management
  - Cleanup of unused resources
  - Progressive loading

## Reviewer notes
Key areas to review:
1. `Visualizer.ts`: Spectrogram rendering logic
2. `WindowFunctions.ts`: Audio processing utilities
3. `ColorMapper.ts`: Color scheme management
4. Performance optimizations in rendering loops
5. Error handling and edge cases

## General notes
The spectrogram visualization feature provides:
- Real-time frequency analysis
- Multiple color schemes for different use cases
- Configurable parameters for detailed analysis
- Smooth integration with existing audio tools
- Optimized performance for large files

robot-ci-heartex committed:
feat(audio): Add spectrogram visualization support to audio editor

This commit adds spectrogram visualization capabilities to the audio editor
through a new optional 'spectrogram' property in the AudioPlus component.

Example usage:
<AudioPlus
  name="audio"
  value="$audio"
  height="240"
  hotkey="space"
  defaultscale="1"
  defaultzoom="2"
  zoom="true"
  spectrogram="true"
  sync="group_a"
/>

Key changes:
- Add new 'spectrogram' boolean property to AudioPlus component
- Extract window functions into a dedicated WindowFunctions module
- Create a new ColorMapper module for spectrogram coloring
- Refactor Visualizer class to use the new modules
- Add support for different window functions and color schemes
- Improve type safety and code organization

The spectrogram visualization allows users to:
- Toggle spectrogram view using the 'spectrogram' property
- View frequency content over time alongside waveform
- Switch between different color schemes
- Configure window functions for FFT analysis
- Adjust visualization parameters (FFT size, dB range)

Configuration:
- spectrogram: boolean (optional) - When set to true, enables
  spectrogram visualization alongside the waveform

Labels: audio, editor, feature, community:feature-request, community:reviewed

Closes #384

cloudmark committed:
feat(audio): Add spectrogram visualization support to audio editor

Add spectrogram visualization capabilities to the audio editor component with configurable settings and improved UI controls.

Key changes:
- Extract window functions into separate WindowFunctions module for better code organization
- Create new ColorMapper module for handling spectrogram color schemes
- Add spectrogram property to AudioPlus component (optional boolean to enable/disable)
- Implement FFT-based spectrogram rendering with configurable parameters
- Add UI controls for spectrogram settings (FFT size, color scheme, dB range)
- Fix CSS styling issues in the configuration modal
- Improve section header positioning and spacing

Features:
- Real-time spectrogram visualization
- Configurable FFT window size and type
- Multiple color scheme options
- Adjustable dB range for visualization
- Mel-scale frequency mapping support
- Responsive rendering with performance optimizations

Labels:
- audio
- community:feature-request
- community:reviewed
- editor
- feature

Closes #384

cloudmark commented:
To help visualize the new spectrogram functionality implemented in this PR (#7400), I've recorded a short video demonstration:

[Video Demonstration: Spectrogram Feature](https://www.loom.com/share/560250b4a9634238b5cb47cb8528ffda?sid=f9b7fb52-22a0-450e-9a70-4506d6001e93)

**What the video shows:**

The video walks through the spectrogram feature within the Label Studio interface, highlighting:
-   The spectrogram display integrated below the audio waveform.
-   Synchronized playback tracking across both the waveform and spectrogram.
-   How zooming affects both views simultaneously.
-   Real-time updates as various configuration options are adjusted in the settings panel:
    -   Changing FFT window sizes.
    -   Applying different color schemes.
    -   Selecting various windowing functions (Blackman, Hann, Hamming).
    -   Toggling the Mel frequency scale.
    -   Adjusting the amplitude (dB) range.

Hope this provides a helpful overview of the user experience!

makseq commented:
Great PR! How well will it work with long audio files around 1-2 hours?

cloudmark commented:
Hey @makseq,

TL; DR: Yes! it handles long files (1-2 hours) efficiently. 

The core strategies implemented are:

1.  **On-Demand Processing:** It only analyzes the audio needed for the *currently visible* portion.
2.  **Efficient Sampling (Zoomed Out):** When zoomed out (e.g., 1 hour on a 1024px view), each pixel represents a time slice (~3.5s). For each pixel, I take one representative FFT (e.g., 512 samples covering ~11.6ms) within that slice, avoiding processing every single audio sample.
3.  **Non-Blocking Rendering:** A generator function (`renderSpectrogramSlice`) renders pixel by pixel, yielding frequently (~16ms) to keep the UI responsive during interactions like scrolling/zooming.

This approach balances performance, memory, and visual overview. As you zoom in, the detail naturally increases as fewer samples are represented per pixel.

Separately, the chosen **FFT window size** affects the computation time *per slice* (larger FFTs = more detail but slower slice render). This characteristic is independent of total file length. For the most fluid feel, 512 is often a good balance.

To demonstrate this with varied audio content, the video uses a 1-hour file created by concatenating samples from the **ESC-50 dataset** ([https://github.com/karolpiczak/ESC-50](https://github.com/karolpiczak/ESC-50)). This dataset contains 2000 short environmental sound recordings across 50 categories (like dogs barking, rain, helicopters, etc.), ensuring the test file has diverse spectral characteristics.

[Video Demo: Spectrogram Performance & FFT Size Impact (1hr ESC-50 file)](https://www.loom.com/share/174ae154ad5045af9e704d65a1103c35?sid=ea661d53-8027-4f50-841e-198bd7e042a3)

*(Video shows loading/panning the long, varied file & the visible speed difference when switching FFT sizes).*

farioas commented:
@cloudmark please rebase your branch on the latest changes from repo to include this commit https://github.com/HumanSignal/label-studio/commit/9b0487f982c7280c7474d228f6d914ab01d1d61a. It will fix failing checks.

cloudmark committed:
Merge branch 'HumanSignal:develop' into feature/spectrogram-analyser

cloudmark commented:
Done @farioas, I also updated https://github.com/HumanSignal/label-studio/pull/7376 to take from this upstream

cloudmark committed:
Merge branch 'HumanSignal:develop' into feature/spectrogram-analyser

makseq commented:
`<AudioPlus>` was deprecated, could you please use `<Audio>` instead?

cloudmark commented:
Thank you @makseq for the heads up.  I think internally they should resolve to the same component so there are no further updates needed (I believe).

nikitabelonogov committed:
Merge remote-tracking branch 'origin/develop' into feature/spectrogram-analyser

bmartel commented on web/libs/editor/src/components/Timeline/Controls/ConfigControl.tsx:
Can we extract the logic introduced here into a separate set of component files, and hooks/utilities, so we don't end up with a 700+ line file in ConfigControl.tsx. We have cursor rules in this project that outline the best practices, aim for 1 component per file where possible, and extract hooks and utils similarly.

bmartel commented on web/libs/editor/src/components/Timeline/Controls/ConfigControl.tsx:
```diff
@@ -1,19 +1,176 @@
+import { createPortal } from "react-dom";
+import {Toggle, Tooltip} from "@humansignal/ui";
+import {Block, Elem} from "../../../utils/bem";
+import {Slider as AntSlider, Select} from "antd";
```
```suggestion
import { Slider as AntSlider } from "antd";
```
I noticed additional usage of AntD components, we are actively removing our dependence on these in the project, so we can use a Select component available from our internal ui lib. See the above suggestion.

bmartel commented on web/libs/editor/src/components/Timeline/Controls/ConfigControl.tsx:
```diff
@@ -1,19 +1,176 @@
+import {Slider as AntSlider, Select} from "antd";
+import {Range} from "../../../common/Range/Range";
+
+import {IconConfig, IconInfoConfig} from "@humansignal/ui";
```
```suggestion
import { IconConfig, IconInfoConfig } from "@humansignal/icons";
```

bmartel commented on web/libs/editor/src/components/Timeline/Controls/ConfigControl.tsx:
```diff
@@ -1,19 +1,176 @@
-import { Slider } from "./Slider";
+import {type FC, type MouseEvent, useContext, useEffect, useMemo, useRef, useState} from "react";
+import { createPortal } from "react-dom";
+import {Toggle, Tooltip} from "@humansignal/ui";
```
```suggestion
import { Select, Toggle, Tooltip } from "@humansignal/ui";
```

bmartel commented:
Hey @cloudmark 👋, can you please run the linter/formatter over your changes as well as rebase there are currently conflicts with a few files.

To lint/fix, from the LabelStudio project root: `make fmt-all`

bmartel commented on web/libs/editor/src/lib/AudioUltra/Visual/Visualizer.ts:
```diff
@@ -139,6 +161,159 @@ export class Visualizer extends Events<VisualizerEvents> {
+
+    // Update Windowing Function
+    if (params.windowingFunction && params.windowingFunction !== this.spectrogramWindowingFunction) {
+      console.log(`Visualizer: Updating Windowing Function to ${params.windowingFunction}`);
```
```suggestion
```

bmartel commented on web/libs/editor/src/lib/AudioUltra/Visual/Visualizer.ts:
```diff
@@ -139,6 +161,159 @@ export class Visualizer extends Events<VisualizerEvents> {
+
+    // Update Color Scheme
+    if (params.colorScheme && params.colorScheme !== this.spectrogramColorScheme) {
+      console.log(`Visualizer: Updating Colors Scheme Function to ${params.colorScheme}`);
```
```suggestion
```

bmartel commented on web/libs/editor/src/lib/AudioUltra/Visual/Visualizer.ts:
```diff
@@ -139,6 +161,159 @@ export class Visualizer extends Events<VisualizerEvents> {
+    // Update dB Range
+    if ((params.minDb !== undefined && params.minDb !== this.spectrogramMinDb) ||
+        (params.maxDb !== undefined && params.maxDb !== this.spectrogramMaxDb)) {
+      console.log(`Visualizer: Updating dB Range to ${params.minDb} - ${params.maxDb}`);
```
```suggestion
```

bmartel commented on web/libs/editor/src/lib/AudioUltra/Visual/Visualizer.ts:
```diff
@@ -139,6 +161,159 @@ export class Visualizer extends Events<VisualizerEvents> {
+    if (needsRedraw) {
+      const spectrogramLayer = this.getLayer("spectrogram");
+      if (spectrogramLayer?.isVisible) {
+        console.log("Visualizer: Redrawing spectrogram due to parameter change.");
```
```suggestion
```

bmartel commented on web/libs/editor/src/lib/AudioUltra/Visual/Visualizer.ts:
```diff
@@ -139,6 +161,159 @@ export class Visualizer extends Events<VisualizerEvents> {
+        spectrogramLayer.clear();
+        this.resetWaveformRender(); // Reset cached render state
+        this.draw(); // Schedule a draw call
+      } else {
+         console.log("Visualizer: Spectrogram layer not visible, skipping redraw.");
```
```suggestion
```

bmartel commented on web/libs/editor/src/lib/AudioUltra/Visual/Visualizer.ts:
```diff
@@ -225,30 +400,63 @@ export class Visualizer extends Events<VisualizerEvents> {
+        } finally {
       this.drawing = false;
-    });
+             // Important: Check if a draw was queued *while* this one was running
+             if (this.isDrawingQueued) {
+                 // We should probably trigger the queued draw now,
+                 // but let's rely on the requestAnimationFrame queue for simplicity first.
+                 // If issues persist, we might need to explicitly call draw() here.
+             }
```
I don't understand what this is required for, seems unused?
```suggestion
```

bmartel commented on web/libs/editor/src/tags/object/AudioUltra/view.tsx:
```diff
@@ -154,6 +188,38 @@ const AudioUltraView: FC<AudioUltraProps> = ({ item }) => {
+
+  const handleFftSamplesChange = useCallback((samples: number) => {
+    setSpectrogramFftSamples(samples);
+    // changeSetting('spectrogramFftSamples', samples); // Persist via context/hook
```
Are the following calls to changeSetting required or not?
```suggestion
```

bmartel commented on web/libs/editor/src/tags/object/AudioUltra/view.tsx:
```diff
@@ -154,6 +188,38 @@ const AudioUltraView: FC<AudioUltraProps> = ({ item }) => {
+
+  const handleMelBandsChange = useCallback((bands: number) => {
+    setNumberOfMelBands(bands);
+    // changeSetting('numberOfMelBands', bands);
```
```suggestion
```

bmartel commented on web/libs/editor/src/tags/object/AudioUltra/view.tsx:
```diff
@@ -154,6 +188,38 @@ const AudioUltraView: FC<AudioUltraProps> = ({ item }) => {
+
+  const handleWindowingFunctionChange = useCallback((windowFunction: string) => {
+    setSpectrogramWindowingFunction(windowFunction);
+    // changeSetting('spectrogramWindowingFunction', windowFunction);
```
```suggestion
```

bmartel commented on web/libs/editor/src/tags/object/AudioUltra/view.tsx:
```diff
@@ -154,6 +188,38 @@ const AudioUltraView: FC<AudioUltraProps> = ({ item }) => {
+
+  const handleColorSchemeChange = useCallback((schemeName: string) => {
+    setSpectrogramColorScheme(schemeName);
+    // changeSetting('spectrogramColorScheme', schemeName);
```
```suggestion
```

bmartel commented on web/libs/editor/src/lib/AudioUltra/hooks/useSpectrogramControls.ts:
```diff
@@ -0,0 +1,134 @@
+    }
+
+    if (Object.keys(paramsToUpdate).length > 0) {
+      console.log("useSpectrogramControls: Updating waveform from context changes", paramsToUpdate);
```
```suggestion
```

bmartel commented on web/libs/editor/src/lib/AudioUltra/Visual/ColorMapper.ts:
```diff
@@ -0,0 +1,136 @@
+        nshades: COLORMAP_NSHADES,
+        format: "float"
+      });
+      console.log(
+        `ColorMapper: Cached colormap '${this.currentScheme}' with ${this.activeColormap.length} shades.`
+      );
```
```suggestion
```

bmartel commented on web/libs/editor/src/lib/AudioUltra/Visual/Visualizer.ts:
```diff
@@ -139,6 +161,159 @@ export class Visualizer extends Events<VisualizerEvents> {
+  private async initFFT() {
+    try {
+      this.fft = new FFT(this.spectrogramFftSamples);
+      console.log(`Visualizer: FFT initialized with size ${this.spectrogramFftSamples}`);
```
```suggestion
```

bmartel commented on web/libs/editor/src/lib/AudioUltra/Visual/Visualizer.ts:
```diff
@@ -139,6 +161,159 @@ export class Visualizer extends Events<VisualizerEvents> {
+
+    // Update FFT Samples
+    if (params.fftSamples && params.fftSamples !== this.spectrogramFftSamples) {
+      console.log(`Visualizer: Updating FFT Samples to ${params.fftSamples}`);
```
```suggestion
```

bmartel commented on web/libs/editor/src/lib/AudioUltra/Visual/Visualizer.ts:
```diff
@@ -139,6 +161,159 @@ export class Visualizer extends Events<VisualizerEvents> {
+
+    // Update Mel Bands
+    if (params.melBands && params.melBands !== this.numberOfMelBands) {
+      console.log(`Visualizer: Updating Mel Bands to ${params.melBands}`);
```
```suggestion
```

bmartel commented on web/libs/editor/src/lib/AudioUltra/Visual/Visualizer.ts:
```diff
@@ -1046,4 +1382,325 @@ export class Visualizer extends Events<VisualizerEvents> {
+
+    // Ensure filterbank is cached or created only when necessary
+    if (!this.melFilterbank || this.melFilterbank.length !== this.numberOfMelBands || this.melFilterbank[0]?.length !== linearBinCount) {
+      console.log(`Creating Mel filterbank: ${this.audio.sampleRate}Hz, ${linearBinCount} linear bins, ${this.numberOfMelBands} mel bands`);
```
```suggestion
```

bmartel commented on web/libs/editor/src/components/Timeline/Controls/ConfigControl.tsx:
```diff
@@ -74,6 +368,190 @@ export const ConfigControl: FC<ConfigControlProps> = ({
+  const [lastUpdate, setLastUpdate] = useState<{min: number; max: number; time: number} | null>(null);
+
+  const handleDbRangeChange = (values: number[]) => {
+    console.log('handleDbRangeChange received:', values);
```
```suggestion
```

bmartel commented on web/libs/editor/src/components/Timeline/Controls/ConfigControl.tsx:
```diff
@@ -74,6 +368,190 @@ export const ConfigControl: FC<ConfigControlProps> = ({
+    console.log('handleDbRangeChange received:', values);
+    
+    if (!Array.isArray(values) || values.length !== 2) {
+      console.log('Invalid values array');
```
```suggestion
```

bmartel commented on web/libs/editor/src/components/Timeline/Controls/ConfigControl.tsx:
```diff
@@ -74,6 +368,190 @@ export const ConfigControl: FC<ConfigControlProps> = ({
+    }
+
+    const [newMinDb, newMaxDb] = values;
+    console.log('Current state before update:', { displayMinDb, displayMaxDb });
+    console.log('New values to set:', { newMinDb, newMaxDb });
```
```suggestion
```

bmartel commented on web/libs/editor/src/components/Timeline/Controls/ConfigControl.tsx:
```diff
@@ -74,6 +368,190 @@ export const ConfigControl: FC<ConfigControlProps> = ({
+
+    // Basic validation
+    if (isNaN(newMinDb) || isNaN(newMaxDb) || newMinDb >= newMaxDb) {
+      console.log('Values invalid or crossed:', { newMinDb, newMaxDb });
```
```suggestion
```

bmartel commented on web/libs/editor/src/components/Timeline/Controls/ConfigControl.tsx:
```diff
@@ -74,6 +368,190 @@ export const ConfigControl: FC<ConfigControlProps> = ({
+    if (lastUpdate && currentTime - lastUpdate.time < 100) {
+      // If we're getting a quick update that would change max when we're moving min
+      if (lastUpdate.min === newMinDb && lastUpdate.max !== newMaxDb && newMaxDb !== displayMaxDb) {
+        console.log('Preventing unstable max update');
```
```suggestion
```

bmartel commented on web/libs/editor/src/components/Timeline/Controls/ConfigControl.tsx:
```diff
@@ -74,6 +368,190 @@ export const ConfigControl: FC<ConfigControlProps> = ({
+      }
+      // If we're getting a quick update that would change min when we're moving max
+      if (lastUpdate.max === newMaxDb && lastUpdate.min !== newMinDb && newMinDb !== displayMinDb) {
+        console.log('Preventing unstable min update');
```
```suggestion
```

bmartel commented on web/libs/editor/src/components/Timeline/Controls/ConfigControl.tsx:
```diff
@@ -74,6 +368,190 @@ export const ConfigControl: FC<ConfigControlProps> = ({
+    // Update local state
+    setDisplayMinDb(newMinDb);
+    setDisplayMaxDb(newMaxDb);
+    console.log('State updated to:', { newMinDb, newMaxDb });
```
```suggestion
```

bmartel commented on web/libs/editor/src/components/Timeline/Controls/ConfigControl.tsx:
```diff
@@ -74,6 +368,190 @@ export const ConfigControl: FC<ConfigControlProps> = ({
+    // Convert string or single number to array if needed
+    const valueArray = Array.isArray(value) ? value : [Number(value)];
+    
+    console.log('Range onChange raw value:', valueArray);
+    console.log('Current display values:', { displayMinDb, displayMaxDb });
```
```suggestion
```

bmartel commented on web/libs/editor/src/components/Timeline/Controls/ConfigControl.tsx:
```diff
@@ -74,6 +368,190 @@ export const ConfigControl: FC<ConfigControlProps> = ({
+    if (!Array.isArray(valueArray) || valueArray.length !== 2) return;
+    
+    let [newMin, newMax] = valueArray.map(Math.round);
+    console.log('After rounding:', { newMin, newMax });
```
```suggestion
```

bmartel commented on web/libs/editor/src/components/Timeline/Controls/ConfigControl.tsx:
```diff
@@ -74,6 +368,190 @@ export const ConfigControl: FC<ConfigControlProps> = ({
+    // Ensure values stay within bounds
+    newMin = Math.max(-120, Math.min(0, newMin));
+    newMax = Math.max(-120, Math.min(0, newMax));
+    console.log('After bounds check:', { newMin, newMax });
```
```suggestion
```

bmartel commented on web/libs/editor/src/components/Timeline/Controls/ConfigControl.tsx:
```diff
@@ -74,6 +368,190 @@ export const ConfigControl: FC<ConfigControlProps> = ({
+
+    // Ensure min is always less than max
+    if (newMin >= newMax) {
+      console.log('Values crossed or equal, adjusting...');
```
```suggestion
```

bmartel commented on web/libs/editor/src/components/Timeline/Controls/ConfigControl.tsx:
```diff
@@ -74,6 +368,190 @@ export const ConfigControl: FC<ConfigControlProps> = ({
+      console.log('Values crossed or equal, adjusting...');
+      if (isMinMoving) {
+        newMin = Math.min(newMin, newMax - 1);
+        console.log('Adjusted min:', newMin);
```
```suggestion
```

bmartel commented on web/libs/editor/src/components/Timeline/Controls/ConfigControl.tsx:
```diff
@@ -74,6 +368,190 @@ export const ConfigControl: FC<ConfigControlProps> = ({
+        console.log('Adjusted min:', newMin);
+      } else {
+        newMax = Math.max(newMax, newMin + 1);
+        console.log('Adjusted max:', newMax);
```
```suggestion
```

bmartel commented on web/libs/editor/src/components/Timeline/Controls/ConfigControl.tsx:
```diff
@@ -74,6 +368,190 @@ export const ConfigControl: FC<ConfigControlProps> = ({
+      }
+    }
+
+    console.log('Final values before handleDbRangeChange:', { newMin, newMax });
```
```suggestion
```

bmartel reviewed (changes requested):
So far other than some linting/code organization changes, this looks amazing 🔥.

cloudmark commented:
Hey @bmartel, I'm going to push an update shortly that speeds up the computations and makes the FFT computations async. LS should be way more responsive than the sync version I presented here.  The render pipeline has also been completely rehauled to optimise rendering.

cloudmark commented:
Here's a short Loom video showing it in action: [https://www.loom.com/share/b2aa14fe5f7e4411811a3437bdc3b729](https://www.google.com/url?sa=E&q=https%3A%2F%2Fwww.loom.com%2Fshare%2Fb2aa14fe5f7e4411811a3437bdc3b729)

Label Studio (LS) should feel much snappier since I moved the FFT computations to be async and overhauled the rendering pipeline.

bmartel committed:
Merge remote-tracking branch 'origin/develop' into feature/spectrogram-analyser

bmartel commented:
@cloudmark Heads up, I resolved the conflicts on this branch and pushed up the changes, as most of it had to do with the recent release of Dark Mode and Design Tokens.

cloudmark committed:
Add spectrogram visualization capabilities to the audio editor component with configurable settings and improved UI controls.

### Key changes:
- Extracted window functions into a separate `WindowFunctions` module for better code organization.
- Created new `ColorMapper` module for handling spectrogram color schemes.
- Added `spectrogram` property to the `AudioPlus` component (optional boolean to enable/disable).
- Implemented FFT-based spectrogram rendering with configurable parameters.
- Added UI controls for spectrogram settings (FFT size, color scheme, dB range).
- Fixed CSS styling issues in the configuration modal.
- Improved section header positioning and spacing.

### Features:
- Real-time spectrogram visualization.
- Configurable FFT window size and type.
- Multiple color scheme options.
- Adjustable dB range for visualization.
- Mel-scale frequency mapping support.
- Responsive rendering with performance optimizations.

### Labels:
- `audio`
- `community:feature-request`
- `community:reviewed`
- `editor`
- `feature`

cloudmark commented:
@bmartel branch updated to include all the updates.

makseq commented:
/git merge develop
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/14918845702)
> Successfully merged: Already up to date.

makseq commented:
/fm sync
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/14918866144)

makseq committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/14918866144

cloudmark committed:
Merge branch 'HumanSignal:feature/spectrogram-analyser' into feature/spectrogram-analyser

cloudmark committed:
feat(audio): improve audio player UI and rendering

- Fix spectrogram visibility by including scrollbar height in rendering pipeline
- Update color scheme to reflect changes immediately
- Optimize progress bar:
  - Reduce size for better UI integration
  - Add auto-hide behavior when no progress is present
- Improve frequency grid layout with optimal Hz label placement

These changes improve the audio player's visual feedback and user experience
by ensuring proper rendering of the spectrogram, making the progress
indicator more subtle and responsive, and enhancing the readability of
frequency labels.

bmartel committed:
Merge remote-tracking branch 'origin/develop' into feature/spectrogram-analyser

bmartel committed:
linting fix scss

bmartel commented on web/libs/editor/src/lib/AudioUltra/Analysis/MelBanks.ts:
```diff
@@ -0,0 +1,163 @@
+      return [];
+    }
+
+    const melPoints = new Float32Array(numBands + 2);
```
```suggestion
    const hzPoints = new Float32Array(numBands + 2);
```

cloudmark replied:
I do not mind the naming, can you do a single commit yourself with the naming you suggested?  If I click on Commit suggestion we will have a number of commits just for this refactor.

bmartel commented on web/libs/editor/src/lib/AudioUltra/Analysis/MelBanks.ts:
```diff
@@ -0,0 +1,163 @@
+
+    // Create evenly spaced points in a Mel scale
+    for (let i = 0; i < numBands + 2; i++) {
+      melPoints[i] = lowFreqMel + i * melStep;
```
Given how this was being used, we can just retain the TypedArray by removing the intermediary step and secondary map call.
```suggestion
      hzPoints[i] = this.melToHz(lowFreqMel + i * melStep);
```

bmartel commented on web/libs/editor/src/lib/AudioUltra/Analysis/MelBanks.ts:
```diff
@@ -0,0 +1,163 @@
+      melPoints[i] = lowFreqMel + i * melStep;
+    }
+
+    const hzPoints: Float32Array = melPoints.map((mel) => this.melToHz(mel));
```
```suggestion
```

bmartel commented on web/libs/editor/src/lib/AudioUltra/Analysis/MelBanks.ts:
```diff
@@ -0,0 +1,163 @@
+    const hzPoints: Float32Array = melPoints.map((mel) => this.melToHz(mel));
+    // Calculate the frequency resolution of the linear FFT bins
+    const freqResolution: number = sampleRate / (2 * (linearBinCount - 1));
+    const binFreqs: Float32Array = new Float32Array(linearBinCount).map((_, i) => i * freqResolution);
```
I'm fairly certain using map on typed arrays can result in a basic array instead of a TypedArray, which is far less performant and does not carry with it the same characteristics.
```suggestion
    const binFreqs: Float32Array = new Float32Array(linearBinCount).map((_, i) => );
    for (let i = 0; i < binFreqs.length; i++) binFreqs[i] = i * freqResolution;
```

bmartel commented on web/libs/editor/src/lib/AudioUltra/Visual/Composition/LayerComposer.ts:
```diff
@@ -0,0 +1,173 @@
+    }
+
+    // Sort layers by z-index (index property) to ensure proper rendering order
+    const layers = [...unsortedLayers].sort((a, b) => a.index - b.index);
```
There is no need to copy the array, the unsortedLayers does not need preservation and this is just overhead.
```suggestion
    const layers = unsortedLayers.sort((a, b) => a.index - b.index);
```

bmartel commented on web/libs/editor/src/lib/AudioUltra/Visual/Composition/LayerComposer.ts:
```diff
@@ -0,0 +1,173 @@
+    }
+
+    // Sort layers by z-index (index property) to ensure proper rendering order
+    const layers = [...unsortedLayers].sort((a, b) => a.index - b.index);
```
```suggestion
    const layers = unsortedLayers.sort((a, b) => a.index - b.index);
```

bmartel commented on web/libs/editor/src/lib/AudioUltra/Visual/Renderer/Plugins/ProgressRendererPlugin.ts:
```diff
@@ -0,0 +1,401 @@
+  }
+}
+
+// Utility function for clamping
+function clamp(value: number, min: number, max: number): number {
+  return Math.max(min, Math.min(max, value));
+}
```
Minor nitpick: Can we move this to a common util file?

bmartel commented on web/libs/editor/src/lib/AudioUltra/Visual/Renderer/SpectrogramRenderer.ts:
```diff
@@ -0,0 +1,731 @@
+      plugin.updateConfig({ visible });
+    }
+
+    // this.resetRenderState();
+    // this.draw(this.lastRenderContext!);
```
Is this not required?
```suggestion
```

bmartel commented on web/libs/editor/src/lib/AudioUltra/Visual/Renderer/WaveformRenderer.ts:
```diff
@@ -0,0 +1,260 @@
+  backgroundColor?: any;
+  middleLineColor?: any;
+  amp?: number;
+  // Add any other configurable options for waveform rendering here
```
```suggestion
```

bmartel commented on web/libs/editor/src/lib/AudioUltra/Visual/Renderer/WaveformRenderer.ts:
```diff
@@ -0,0 +1,260 @@
+        while (i > 0) {
+          const index = l - i;
+          const chunk = chunks.slice(index, index + context.samplesPerPx);
+          if (now - performance.now() > 10) {
```
I'm fairly certain this would have never yielded previously since the values would have been negative.
```suggestion
          if (performance.now() - now > 10) {
```

Looking at it further, the value of now should be reset after it yields to performance.now() otherwise it won't ever yield again if needed.

bmartel commented on web/libs/editor/src/tags/control/Taxonomy/Taxonomy.scss:
```diff
@@ -81,4 +81,8 @@ body :global(.ant-select-multiple .ant-select-selection-item-remove) {
 
 body :global(.ant-select-arrow) {
   color: var(--color-neutral-icon);
+}
+
+body :global(.ant-select-single .ant-select-selection-item) {
+  color: var(--color-neutral-content);
```
Why is a change to the Taxonomy tag required for this PR?

cloudmark replied:
I think so yes, without that the selection items look incorrect in  dark mode

bmartel commented on web/libs/editor/src/tags/object/AudioUltra/view.tsx:
```diff
@@ -156,6 +195,53 @@ const AudioUltraView: FC<AudioUltraProps> = ({ item }) => {
+  const handleFftSamplesChange = useCallback(
+    (samples: number) => {
+      setSpectrogramFftSamples(samples);
+      // changeSetting('spectrogramFftSamples', samples); // Persist via context/hook
```
```suggestion
```

bmartel commented on web/libs/editor/src/tags/object/AudioUltra/view.tsx:
```diff
@@ -156,6 +195,53 @@ const AudioUltraView: FC<AudioUltraProps> = ({ item }) => {
+  const handleMelBandsChange = useCallback(
+    (bands: number) => {
+      setNumberOfMelBands(bands);
+      // changeSetting('numberOfMelBands', bands);
```
```suggestion
```

bmartel commented on web/libs/editor/src/tags/object/AudioUltra/view.tsx:
```diff
@@ -156,6 +195,53 @@ const AudioUltraView: FC<AudioUltraProps> = ({ item }) => {
+  const handleWindowingFunctionChange = useCallback(
+    (windowFunction: WindowFunctionType) => {
+      setSpectrogramWindowingFunction(windowFunction);
+      // changeSetting('spectrogramWindowingFunction', windowFunction);
```
```suggestion
```

bmartel commented on web/libs/editor/src/tags/object/AudioUltra/view.tsx:
```diff
@@ -156,6 +195,53 @@ const AudioUltraView: FC<AudioUltraProps> = ({ item }) => {
+  const handleColorSchemeChange = useCallback(
+    (schemeName: ColorScheme) => {
+      setSpectrogramColorScheme(schemeName);
+      // changeSetting('spectrogramColorScheme', schemeName);
```
```suggestion
```

bmartel committed:
Merge remote-tracking branch 'origin/develop' into feature/spectrogram-analyser

bmartel committed:
remove antd from spectrogram control

cloudmark committed:
fix(audio): improve spectrogram performance and controls

- Add rate limiting to spectrogram rendering to cap computation frequency
- Fix resize handler to properly update spectrogram dimensions
- Fix spectrogram controls to properly handle input visibility
- Optimize zoom performance with debounced updates

This change improves the overall performance and stability of the spectrogram
visualization by preventing excessive computations and ensuring proper
dimension updates during resize events.

cloudmark committed:
fix: resolve waveform rendering artifacts and maintain backward compatibility

- Fix waveform seam artifacts by:
  - Implementing forced full redraw to avoid anti-aliasing issues
  - Adding proper seam gap handling with WAVEFORM_SEAM_GAP_FILL
  - Ensuring pixel-ratio safe coordinate calculations
  - Moving seam-related constants to central constants file

- Maintain backward compatibility:
  - Revert to showing full waveform on initial load
  - Remove 10-second view limitation
  - Preserve existing zoom behavior

- Enhance rendering architecture:
  - Add public pixelRatio getter to Layer class
  - Improve type safety and property access patterns
  - Centralize rendering constants
  - Fix linter errors and improve code organization

These changes resolve the visual artifacts caused by HTML5 Canvas's
anti-aliasing of vector graphics while maintaining compatibility with
existing behavior. The code is now more maintainable with proper
encapsulation and centralized constants.

bmartel committed:
Merge remote-tracking branch 'origin/develop' into feature/spectrogram-analyser

cloudmark committed:
fix: prevent timeline label overlap when fully zoomed out

- Enforce a minimum pixel spacing between timeline labels (default 40px)
- Only render a label if it is far enough from the previous one
- Ensures labels remain readable and do not overlap at any zoom level

bmartel committed:
revert the pixelRatio getter and property rename

cloudmark committed:
chore: removed optimisations to the waveform, we will address these once the canvas API provides support for anti-aliasing.

bmartel committed:
ensure the changes made to rendering are behind the FF

bmartel committed:
target 60fps for all renderers baseline, make spectrogram renderer target 15fps to maintain performance of application

bmartel commented:
/git merge
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/15443778059)
> Successfully merged:  create mode 100644 web/libs/editor/src/components/TaskSummary/types.ts

bmartel committed:
Merge branch 'develop' into 'feature/spectrogram-analyser'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/15443778059

bmartel commented:
/fmt
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/15443883051)

bmartel committed:
fixing lint errors

bmartel committed:
running linter over changed files in AudioUltra

bmartel commented on web/libs/editor/src/lib/AudioUltra/Visual/Renderer/Plugins/GridRendererPlugin.ts:
```diff
@@ -154,8 +154,8 @@ export class GridRendererPlugin implements RendererPlugin<GridRendererPluginConf
       const mel = (f: number) => 2595 * Math.log10(1 + f / 700);
-      const invMel = (m: number) => 700 * (Math.pow(10, m / 2595) - 1);
-      const melMax = mel(nyquist);
+      const _invMel = (m: number) => 700 * (10 ** (m / 2595) - 1);
+      const _melMax = mel(nyquist);
```
@cloudmark these changes don't appear used anywhere, hence why the linter put a `_` before them. Can we double check and clean up this aspect? Also, note that removing these will likely also make other values unused and so on, I guess I just question whether this code path is doing what is expected.

bmartel commented on web/libs/editor/src/lib/AudioUltra/Visual/Renderer/Plugins/ProgressRendererPlugin.ts:
```diff
@@ -191,10 +195,10 @@ export class ProgressRendererPlugin implements RendererPlugin<ProgressRendererPl
       if (fftCache) {
         for (const channelMap of fftCache.values()) {
-          fftCacheCount += channelMap.size;
+          _fftCacheCount += channelMap.size;
```
@cloudmark This value is unused, can we remove it and what are the impacts if we do?

bmartel commented on web/libs/editor/src/lib/AudioUltra/Visual/Renderer/Plugins/ProgressRendererPlugin.ts:
```diff
@@ -223,7 +227,7 @@ export class ProgressRendererPlugin implements RendererPlugin<ProgressRendererPl
    */
   private _createPriorityBars(batchGroups: Record<string, BatchProgress[]>): HTMLElement {
-    const cfg = this.config;
+    const _cfg = this.config;
```
@cloudmark Another unused value to either remove or adapt and keep.

bmartel commented on web/libs/editor/src/lib/AudioUltra/Visual/Renderer/SpectrogramRenderer.ts:
```diff
@@ -532,7 +531,7 @@ export class SpectrogramRenderer implements Renderer<SpectrogramRendererConfig>
     }
     const fftSamples = this.fftProcessor.fftSamples;
-    const baseHopSize = Math.max(1, Math.floor(fftSamples / this.spectrogramHopFactor));
+    const _baseHopSize = Math.max(1, Math.floor(fftSamples / this.spectrogramHopFactor));
```
@cloudmark Another unused value to either remove or adapt and keep.

bmartel committed:
Merge remote-tracking branch 'origin/develop' into feature/spectrogram-analyser

cloudmark committed:
chore: removed unused variables / values

cloudmark commented:
All done @bmartel, thank you

bmartel committed:
fix issue with FFT being compiled on the main thread, moved to a worker

bmartel committed:
Merge remote-tracking branch 'origin/develop' into feature/spectrogram-analyser

bmartel committed:
Revert "fix issue with FFT being compiled on the main thread, moved to a worker"

This reverts commit 8d3c6ad51eed2d4a583f7f23d2f767c9c97e8bce.

bmartel committed:
fix tests from loading and compiling fft on the main thread

bmartel committed:
fix audio control tests

bmartel committed:
fix visualizer render to ensure all renderers are redrawn before compisiting happens

bmartel committed:
fix playhead render not being correct on zoom in/out

bmartel committed:
fix integration test for audio relations

bmartel committed:
fix zooming when no playback

bmartel committed:
allow audio config modal headers to be sticky on scroll

bmartel committed:
removing the matrix ff tests, as we only have one actual case now

bmartel committed:
removing the matrix ff tests, as we only have one actual case now

bmartel committed:
need to add a few more ticks to waits to ensure it passes reliably on CI

bmartel committed:
fix flakey CI tests for Audio/Video/Paragraphs sync

bmartel committed:
change wait times for av sync tests

bmartel committed:
removing test, converted on parent branch
